scipy/stats/_morestats.py

```diff
@@ -1311,28 +1336,34 @@ def _all(x):
         message = ("The `optimizer` argument of `boxcox_normmax` must return "
                    "an object containing the optimal `lmbda` in attribute `x`.")
         raise ValueError(message)
-    else:
-        # Test if the optimal lambda causes overflow
-        x = np.asarray(x)
-        x_treme = np.max(x, axis=0) if np.any(res > 0) else np.min(x, axis=0)
-        istransinf = np.isinf(special.boxcox(x_treme, res))
-        dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
-        if np.any(istransinf):
-            warnings.warn(
+    elif not np.isinf(ymax):  # adjust the final lambda
+        # x > 1, boxcox(x) > 0; x < 1, boxcox(x) < 0
+        xmax, xmin = np.max(x), np.min(x)
+        if xmin >= 1:
+            x_treme = xmax
+        elif xmax <= 1:
+            x_treme = xmin
+        else:  # xmin < 1 < xmax
+            indicator = special.boxcox(xmax, res) > abs(special.boxcox(xmin, res))
+            if isinstance(res, np.ndarray):
+                indicator = indicator[1]  # select corresponds with 'mle'
+            x_treme = xmax if indicator else xmin
+
+        mask = abs(special.boxcox(x_treme, res)) > ymax
+        if np.any(mask):
+            message = (
                 f"The optimal lambda is {res}, but the returned lambda is the"
                 f"constrained optimum to ensure that the maximum or the minimum "
-                f"of the transformed data does not cause overflow in {dtype}.",
-                stacklevel=2
+                f"of the transformed data does not " + end_msg
             )
+            warnings.warn(message, stacklevel=2)
 
             # Return the constrained lambda to ensure the transformation
-            # does not cause overflow. 10000 is a safety factor because
-            # `special.boxcox` overflows prematurely.
-            ymax = np.finfo(dtype).max / 10000
-            constrained_res = _boxcox_inv_lmbda(x_treme, ymax * np.sign(res))
+            # does not cause overflow or exceed specified `ymax`
+            constrained_res = _boxcox_inv_lmbda(x_treme, ymax * np.sign(x_treme - 1))
```

Comment:
Assuming that `_boxcox_inv_lmbda` is vectorised, couldn't this be simplified to something like:
```python
elif not np.isinf(ymax):  # adjust the final lambda
    constrained_res = min(
        np.min(_boxcox_inv_lmbda(x, -ymax)),
        np.min(_boxcox_inv_lmbda(x, ymax))
    )
```

Reply:
1.  I think the code is not equivalent. I test with the data that cause negative overflow.
```python
x = np.array([0.50000471, 0.50004979, 0.50005902, 0.50009312, 0.50001632])
```
2. Compute the inverse on the full data is not efficient. We only need to consider `x_treme`.

Reply:
Apologies for not being clear: that bit of code was not intended as a drop-in replacement, but rather as pseudo-code to share the idea that most of that logic (and hence maintenance burden) could potentially be avoided. Here is a second attempt that is closer to a drop-in replacement:
```python
elif not np.isinf(ymax):  # adjust the final lambda
    constrained_res = np.clip(
        res,
        np.max(_boxcox_inv_lmbda(x[x < 1], -ymax), initial=-np.inf),
        np.min(_boxcox_inv_lmbda(x[x > 1], ymax), initial=np.inf)
    )
```

In terms of efficiency I don't think this change would materially impact the compute time as this is more or less equivalent to just one additional function evaluation.

Reply:
That is less efficient, so it's reasonable that it's not what we have already. That said, there is a balance between efficiency and complexity, and I prefer the suggestion.

I take it that instead of clipping, you'd prefer bounding the optimization to these values?

Since the transformation is monotonic in `x`, though, I don't see why not take `max/min` before inversion.